framework/include/linesearches/ContactLineSearch.h

```diff
@@ -0,0 +1,85 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef CONTACTLINESEARCH_H
+#define CONTACTLINESEARCH_H
+
+#include "LineSearch.h"
+
+/**
+ * This class implements a custom line search for use with
+ * mechanical contact. The line search is not fancy. It takes two parameters, set in the MOOSE
+ * Executioner block: `contact_line_search_ltol` and `contact_line_search_allowed_lambda_cuts`. The
+ * allowed_lambda_cuts parameter specifies the number of times the line search is allowed to cut
+ * lambda. If allowed to be cut, lambda will be reduced by half, and a new residual will be
+ * evaluated. If the residual is smaller with a smaller lambda, then cuts will continue until
+ * reaching allowed_lambda_cuts. If the residual is larger with a smaller lambda, then the line
+ * search is curtailed and the smaller residual is used. It's recommended that allowed_lambda_cuts
+ * be <= 3, with smaller values being used for smaller contact problems. This is to allow necessary
+ * residual increases when the transient problem requires significant changes in the contact state.
+ *
+ * When the contact set is changing, the user may optionally use a looser linear tolerance set by
+ * the `contact_line_search_ltol` parameter. Then when the contact set is changing during the
+ * beginning of the Newton solve, unnecessary computational expense is avoided. Then when the
+ * contact set is resolved late in the Newton solve, the linear tolerance will return to the finer
+ * tolerance set through the traditional `l_tol` parameter.
+ */
+class ContactLineSearch : public LineSearch
+{
+public:
+  ContactLineSearch(FEProblemBase & fe_problem,
+                    MooseApp & app,
+                    size_t allowed_lambda_cuts,
+                    Real contact_ltol,
+                    bool affect_ltol);
+
+  // A dummy pure virtual destructor because this must be a base class; the derived class must
+  // implement the line searching capabilities
+  virtual ~ContactLineSearch() = 0;
```

Comment:
... but this isn't a base class. It directly inherits from LineSearch which also has a pure virtual destructor. Neither of those make any sense.

Reply:
The point of these was to ensure that no one could create instances of these base classes because they are incomplete with respect to functionality. I agree this is not a great way to do this.

Ideally, there should be a virtual `linesearch` method in the base class that the user needs to fill out, as you suggested in your summary review. I don't know how to make this work, however, with the current `ComputeLineSearchObject` in libmesh that looks like this:
```c++
  class ComputeLineSearchObject
  {
  public:
    virtual ~ComputeLineSearchObject () {}

    virtual void linesearch (SNESLineSearch linesearch) = 0;
  };
```
I think that something like this should be the base class, but what should the `linesearch` signature look like? It obviously can't take a `SNESLineSearch` always as an argument because we want to abstract this from Petsc.

Reply:
Well the MOOSE-like way wouldn't pass in the object anyway. We'd just have a linesearch method. Instead we'd have a shim class that would encapsulate the linesearch object to act on, but it'd be a member of the class. We should be able to come up with an abstraction that works cleanly everywhere and people can just dial up. This would be much like the ParsedFunction object or the libMeshPartitioner object. When the system is half-way between MOOSE and libMesh, there's a little bit of work to come up with a clean abstraction. Let's chat with @friedmud about the design when he gets in.